kuma/core/middleware.py

```diff
@@ -53,7 +54,7 @@ def process_request(self, request):
             return response
 
         request.path_info = '/' + prefixer.shortened_path
-        request.locale = prefixer.locale
+        request.locale = prefixer.locale or settings.LANGUAGE_CODE
```

Comment:
That's really the issue of the custom URL prefixer and middleware we use. It confuses URL handling with language detection and completely works around Django's accepted behavior of taking `Accept-Locale` into account. It's not right to say that any non-locale specific URL doesn't require a locale attached to the request, since *any* request happens under some kind of environment that has a language. There is simply not a "language free" environment from where requests are made from. Consequently we must make sure `request.LANGUAGE_CODE` points at what we consider to be the correct language for the request, falling back to the default language if we can't deduce it from the URL.

So I think you're right we should do 1) *and* 2). The first to cover all angles and follow the idea of any request having some language and the latter to match Django's behavior of using "LANGUAGE_CODE" to define the request's language code.